DIPs/DIP1xxx-rn.md

```diff
@@ -0,0 +1,927 @@
+
+### Syntax
+
+The following link exhibits the proposed syntax and the necessary grammar changes in order to support
+the copy construct:
+
+https://github.com/dlang/dlang.org/pull/2414
+
+A declaration is a copy constructor declaration if it is a constructor declaration annotated with
+the `@implicit` attribute and takes only one parameter by reference that is of the same type as
+`typeof(this)`. `@implicit` is a compiler recognised attribute like `@safe` or `nogc` and it can be
+syntactally used the same as the others. `@implicit` can be legally used solely to mark the declaration
+of a copy constructor; all other uses of this attribute will result in a compiler error. The proposed syntax
+benefits from the advantage of declaring the copy constructor in an expressive manner without adding
+additional complexity to the existing syntax.
+
+The copy constructor needs to be annotated with `@implicit` in order to distinguish a copy
+constructor from a normal constructor and avoid silent modification of code behavior. Consider
+this example:
+
+```d
+import std.stdio;
+struct A {
+    this(ref immutable A obj) { writeln("x"); }
+}
+void main()
+{
+    immutable A ia;
+    A a = A(ia);
+    A b = ia;
+}
+```
+
+With the current state of the language, `x` is printed once. with the addition of the copy constructor without
+`@implicit`, `x` would be printed twice, thus modifying the semantics of existing code.
+
+The argument to the copy constructor is passed by reference in order to avoid infinite recursion (passing by
+value would require a copy of the `struct` which would be made by calling the copy constructor, thus leading
+to an infinite chain of calls).
+
+The type qualifiers may be applied to the parameter of the copy constructor, but also to the function itself
+in order to facilitate the ability to describe copies between objects of different mutability levels. The type
+qualifiers are optional.
+
+### Semantics
+
+This sections discusses all the aspects regarding the semantic analysis and interaction with other
+components of the copy constructor
+
+#### Copy constructor usage
+
+The copy constructor is implicitly used by the compiler whenever a struct variable is initialized
+as a copy of another variable of the same unqualified type:
+
+1. When a variable is explicitly initialized:
+
+```d
+struct A
+{
+    @implicit this(ref A another) {}
+}
+
+void main()
+{
+    A a;
+    A b = a; // copy constructor gets called
+    b = a;   // assignment, not initialization
+}
+```
+
+2. When a parameter is passed by value to a function:
+
+```d
+struct A
+{
+    @implicit this(ref A another) {}
+}
+
+void fun(A a) {}
+
+void main()
+{
+    A a;
+    fun(a);    // copy constructor gets called
+}
+```
+
+3. When a parameter is returned by value from a function:
+
+```d
+struct A
+{
+    @implicit this(ref A another) {}
+}
+
+A fun()
+{
+    A a;
+    return a;
+}
+
+void main()
+{
+    A a;
+    a = fun();      // copy constructor gets called
```

Comment:
Wouldn't this be a breaking change to require the copy constructor in this case? Right now move semantics are used, so this can be done without a postblit constructor `@disable this(this);`. Otherwise what rules are going to be used to determine whether a move can simply be done a copy constructor needs to be called?

```d
struct A
{
    @disable this(this);
    
    int a;
}

A func()
{
    return A(0);
}

void main()
{
    A a;
    a = func(); // ok
}
```

Reply:
You are right, the value is moved, but in my example the function returns an lvalue and a call to the copy constructor is issued on callee site. This is explained in the next paragraph.

Reply:
Yes but why would it do a Copy if it can just do a Move instead? In this instance when the copy constructor is the same type.

Reply:
The copy constructor may be used for conversions:

```d
struct A
{    
    int[] a;
    @implicit this(ref immutable A rhs) {}
}

A func(immutable A a)
{
    return a;
}

void main()
{
    A a;
    a = func();
}
```